File: lib/parseImageMetadata.ts
/**
 * Image Metadata Parser
 *
 * Extracts AI generation metadata from images (PNG, JPEG)
 * Supports: A1111/Civitai, ComfyUI, NovelAI formats
 */

interface ParsedMetadata {
  format: 'A1111/Civitai' | 'ComfyUI' | 'NovelAI' | 'A1111/Civitai (JPEG)' | 'ComfyUI (JPEG)';
  params: Record<string, string>;
  raw_text: string;
  workflow_text?: string;
}

/**
 * Parse A1111/Automatic1111/Civitai format metadata
 * Format: "prompt text\nNegative prompt: negative text\nSteps: 20, Sampler: Euler a, CFG scale: 7, ..."
 */
function parseA1111Metadata(text: string): Record<string, string> {
  const params: Record<string, string> = {};

  try {
    const parts = text.split('Steps: ');
    if (parts.length !== 2) {
      throw new Error('Invalid A1111 format: missing Steps');
    }

    const prompts = parts[0];
    const parameters = 'Steps: ' + parts[1];

    // Extract positive and negative prompts
    if (prompts.includes('Negative prompt: ')) {
      const promptParts = prompts.split('Negative prompt: ');
      params['Prompt'] = promptParts[0].trim();
      params['Negative Prompt'] = promptParts[1].trim();
    } else {
      params['Prompt'] = prompts.trim();
    }

    // Extract parameters (Steps, Sampler, CFG, etc.)
    const paramList = parameters.split(', ');
    for (const param of paramList) {
      const [key, ...valueParts] = param.split(': ');
      if (key && valueParts.length > 0) {
        params[key.trim()] = valueParts.join(': ').trim();
      }
    }
  } catch (error) {
    params['Error'] = `Failed to parse: ${error}`;
  }

  return params;
}

/**
 * Parse ComfyUI workflow metadata
 * Format: JSON with node IDs and their inputs
 */
function parseComfyUIMetadata(jsonText: string): Record<string, string> {
  const params: Record<string, string> = {};

  try {
    const data = JSON.parse(jsonText);

    // Extract key information from nodes
    for (const [nodeId, nodeData] of Object.entries(data)) {
      const node = nodeData as any;
      const inputs = node.inputs || {};
      const classType = node.class_type || '';

      // Extract text prompts
      if (inputs.text && typeof inputs.text === 'string') {
        params[`${classType} (${nodeId})`] = inputs.text;
      }

      // Extract checkpoint/model
      if (inputs.ckpt_name) {
        params[`Checkpoint (${nodeId})`] = inputs.ckpt_name;
      }

      // Extract common parameters
      if (inputs.seed || inputs.noise_seed) {
        params[`Seed (${nodeId})`] = String(inputs.seed || inputs.noise_seed);
      }
      if (inputs.steps) {
        params[`Steps (${nodeId})`] = String(inputs.steps);
      }
      if (inputs.cfg) {
        params[`CFG (${nodeId})`] = String(inputs.cfg);
      }
    }
  } catch (error) {
    params['Error'] = `Failed to parse ComfyUI JSON: ${error}`;
  }

  return params;
}

/**
 * Parse NovelAI metadata
 * Format: JSON in Comment field
 */
function parseNovelAIMetadata(jsonText: string): Record<string, string> {
  const params: Record<string, string> = {};

  try {
    const data = JSON.parse(jsonText);

    params['Prompt'] = data.prompt || '';
    params['Negative Prompt'] = data.uc || '';
    params['Steps'] = String(data.steps || '');
    params['Sampler'] = data.sampler || '';
    params['CFG scale'] = String(data.scale || '');
    params['Seed'] = String(data.seed || '');

    if (data.width && data.height) {
      params['Size'] = `${data.width}x${data.height}`;
    }
    if (data.Source) {
      params['Model'] = data.Source;
    }
  } catch (error) {
    params['Error'] = `Failed to parse NovelAI JSON: ${error}`;
  }

  return params;
}

/**
 * Read PNG text chunks using browser APIs
 */
async function readPNGMetadata(arrayBuffer: ArrayBuffer): Promise<ParsedMetadata | null> {
  const bytes = new Uint8Array(arrayBuffer);

  // PNG signature check
  if (bytes[0] !== 0x89 || bytes[1] !== 0x50 || bytes[2] !== 0x4E || bytes[3] !== 0x47) {
    return null;
  }

  const decoder = new TextDecoder('utf-8');
  let offset = 8; // Skip PNG signature

  const textChunks: Record<string, string> = {};

  // Read chunks
  while (offset < bytes.length) {
    const length = new DataView(bytes.buffer, offset, 4).getUint32(0);
    const type = decoder.decode(bytes.slice(offset + 4, offset + 8));
    const data = bytes.slice(offset + 8, offset + 8 + length);

    if (type === 'tEXt' || type === 'zTXt' || type === 'iTXt') {
      // Find null separator
      let nullIndex = 0;
      for (let i = 0; i < data.length; i++) {
        if (data[i] === 0) {
          nullIndex = i;
          break;
        }
      }

      const keyword = decoder.decode(data.slice(0, nullIndex));
      const value = decoder.decode(data.slice(nullIndex + 1));
      textChunks[keyword] = value;
    }

    offset += 12 + length; // length(4) + type(4) + data(length) + crc(4)

    if (type === 'IEND') break;
  }

  // Check for NovelAI
  if (textChunks['Software'] === 'NovelAI' && textChunks['Comment']) {
    return {
      format: 'NovelAI',
      params: parseNovelAIMetadata(textChunks['Comment']),
      raw_text: textChunks['Comment']
    };
  }

  // Check for A1111
  if (textChunks['parameters'] && textChunks['parameters'].includes('Steps:')) {
    return {
      format: 'A1111/Civitai',
      params: parseA1111Metadata(textChunks['parameters']),
      raw_text: textChunks['parameters']
    };
  }

  // Check for ComfyUI
  if (textChunks['prompt'] && textChunks['prompt'].trim().startsWith('{')) {
    return {
      format: 'ComfyUI',
      params: parseComfyUIMetadata(textChunks['prompt']),
      raw_text: textChunks['prompt'],
      workflow_text: textChunks['workflow']
    };
  }

  return null;
}

/**
 * Read JPEG EXIF data (for Civitai images)
 * Note: This is a simplified version - for production use exifr library
 */
async function readJPEGMetadata(arrayBuffer: ArrayBuffer): Promise<ParsedMetadata | null> {
  // For proper JPEG EXIF parsing, you should use the 'exifr' library
  // This is a basic implementation that looks for Civitai's UserComment field

  const bytes = new Uint8Array(arrayBuffer);

  // JPEG signature check
  if (bytes[0] !== 0xFF || bytes[1] !== 0xD8) {
    return null;
  }

  // Look for APP1 marker (EXIF)
  let offset = 2;
  while (offset < bytes.length - 1) {
    if (bytes[offset] === 0xFF && bytes[offset + 1] === 0xE1) {
      // Found APP1
      const length = (bytes[offset + 2] << 8) | bytes[offset + 3];
      const exifData = bytes.slice(offset + 4, offset + 2 + length);

      // Look for "UNICODE" prefix in the data (Civitai's UserComment format)
      const decoder = new TextDecoder('utf-8');
      const exifText = decoder.decode(exifData);

      if (exifText.includes('UNICODE')) {
        // Find the UNICODE marker
        const unicodeIndex = exifText.indexOf('UNICODE');
        if (unicodeIndex !== -1) {
          // Extract bytes after UNICODE\x00\x00
          const commentStart = offset + 4 + unicodeIndex + 8;
          let commentBytes = bytes.slice(commentStart, offset + 2 + length);

          // Strip null bytes (Civitai stores as UTF-16-LE with nulls)
          const strippedBytes = commentBytes.filter(b => b !== 0);
          const comment = decoder.decode(strippedBytes);

          // Check if it's ComfyUI JSON or A1111 text
          if (comment.trim().startsWith('{')) {
            return {
              format: 'ComfyUI (JPEG)',
              params: parseComfyUIMetadata(comment),
              raw_text: comment
            };
          } else if (comment.includes('Steps:')) {
            return {
              format: 'A1111/Civitai (JPEG)',
              params: parseA1111Metadata(comment),
              raw_text: comment
            };
          }
        }
      }
      break;
    }
    offset++;
  }

  return null;
}

/**
 * Main function to parse image metadata
 * Supports both PNG and JPEG formats
 */
export async function parseImageMetadata(file: File): Promise<ParsedMetadata | null> {
  const arrayBuffer = await file.arrayBuffer();
  const bytes = new Uint8Array(arrayBuffer);

  // Detect file type
  const isPNG = bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4E && bytes[3] === 0x47;
  const isJPEG = bytes[0] === 0xFF && bytes[1] === 0xD8;

  if (isPNG) {
    return readPNGMetadata(arrayBuffer);
  } else if (isJPEG) {
    return readJPEGMetadata(arrayBuffer);
  }

  return null;
}

/**
 * Format metadata for display
 */
export function formatMetadataForDisplay(metadata: ParsedMetadata): string {
  let output = `Format: ${metadata.format}\n\n`;
  output += 'Parameters:\n';
  output += '='.repeat(80) + '\n\n';

  for (const [key, value] of Object.entries(metadata.params)) {
    output += `${key}:\n  ${value}\n\n`;
  }

  return output;
}

File: parse_image_metadata.py
#!/usr/bin/env python3
"""
Simplified image metadata parser for A1111, ComfyUI, and NovelAI formats.
Extracted from PromptInspectorBot - works with PNG and JPEG (though JPEG often strips metadata).
"""

import io
import json
import re
import sys
from collections import OrderedDict
from pathlib import Path
from PIL import Image


def sanitize_text(text, max_length=10000):
    """Sanitize text content to only allow specific characters"""
    if not isinstance(text, str):
        return ""

    text = re.sub(r"https?://\S+|ftp://\S+|www\.\S+", "", text)
    text = text[:max_length]
    text = re.sub(r'[^A-Za-z0-9\(\)_\-<>:,\{\}\'"\ \n\r\\\[\]\.\|]', "", text)
    return text


def safe_json_loads(json_str, default=None):
    """Safely parse JSON with limits on size and recursion"""
    if not isinstance(json_str, str) or len(json_str) > 1_000_000:
        return default
    try:
        return json.loads(json_str)
    except (json.JSONDecodeError, RecursionError, MemoryError):
        return default


class MetadataA1111:
    """Parse Automatic1111/Civitai format metadata"""
    NAME = "A1111"

    def __init__(self, param_str):
        self.raw_text = param_str
        self.params = self.parse(param_str)

    def parse(self, param_str: str) -> OrderedDict:
        param_str = sanitize_text(param_str, 50000)
        output_dict = OrderedDict()

        try:
            parts = param_str.split("Steps: ", 1)
            if len(parts) != 2:
                raise ValueError("Can't parse A1111 metadata: missing Steps key")

            prompts = parts[0]
            params = "Steps: " + parts[1]
            neg_parts = prompts.split("Negative prompt: ", 1) if "Negative prompt: " in prompts else ()

            if neg_parts:
                output_dict["Prompt"] = sanitize_text(neg_parts[0].strip(), 1000)
                output_dict["Negative Prompt"] = sanitize_text(neg_parts[1].strip(), 1000)
            else:
                output_dict["Prompt"] = sanitize_text(prompts.strip(), 1000)

            params = params.split(", ")
            for param in params:
                parts = param.split(": ", 1)
                if len(parts) == 2:
                    key = sanitize_text(parts[0].strip(), 100)
                    value = sanitize_text(parts[1].strip(), 1000)
                    output_dict[key] = value

        except Exception as e:
            output_dict["Error"] = f"Could not parse metadata: {e}"

        return output_dict


class MetadataComfyUI:
    """Parse ComfyUI format metadata"""
    NAME = "ComfyUI"

    def __init__(self, prompt_str, workflow_str=None):
        self.raw_text = prompt_str
        self.workflow_text = workflow_str
        self.params = self.parse(prompt_str, workflow_str)

    def parse(self, prompt_str: str, workflow_str: str = None) -> OrderedDict:
        promptdata = safe_json_loads(prompt_str, {})
        workflowdata = safe_json_loads(workflow_str, {}) if workflow_str else {}

        output_dict = OrderedDict()

        # Extract key information from ComfyUI nodes
        for node_id, node_data in promptdata.items():
            try:
                inputs = node_data.get("inputs", {})
                class_type = node_data.get("class_type", "")

                # Extract common useful fields
                if "text" in inputs:
                    key = f"{class_type} ({node_id})"
                    output_dict[key] = sanitize_text(str(inputs["text"]), 1000)

                if "ckpt_name" in inputs:
                    output_dict[f"Checkpoint ({node_id})"] = inputs["ckpt_name"]

                if "seed" in inputs or "noise_seed" in inputs:
                    seed = inputs.get("seed") or inputs.get("noise_seed")
                    output_dict[f"Seed ({node_id})"] = str(seed)

                if "steps" in inputs:
                    output_dict[f"Steps ({node_id})"] = str(inputs["steps"])

                if "cfg" in inputs:
                    output_dict[f"CFG ({node_id})"] = str(inputs["cfg"])

            except Exception:
                pass

        return output_dict


class MetadataNovelAI:
    """Parse NovelAI format metadata"""
    NAME = "NovelAI"

    def __init__(self, comment_str):
        self.raw_text = comment_str
        self.params = self.parse(comment_str)

    def parse(self, param_str: str) -> OrderedDict:
        param_str = sanitize_text(param_str, 50000)
        output_dict = OrderedDict()

        try:
            data = safe_json_loads(param_str, {})

            # Extract prompts
            output_dict["Prompt"] = sanitize_text(data.get("prompt", ""), 1000)
            output_dict["Negative Prompt"] = sanitize_text(data.get("uc", ""), 1000)

            # Extract parameters
            output_dict["Steps"] = str(data.get("steps", ""))
            output_dict["Sampler"] = sanitize_text(data.get("sampler", ""), 100)
            output_dict["CFG scale"] = str(data.get("scale", ""))
            output_dict["Seed"] = str(data.get("seed", ""))

            width = data.get("width", "")
            height = data.get("height", "")
            if width and height:
                output_dict["Size"] = f"{width}x{height}"

            if "Source" in data:
                output_dict["Model"] = sanitize_text(data.get("Source", ""), 100)

        except Exception as e:
            output_dict["Error"] = f"Could not parse metadata: {e}"

        return output_dict


def extract_metadata(image_path: str) -> dict:
    """
    Extract metadata from an image file.

    Returns:
        dict with keys: 'format' (str), 'params' (OrderedDict), 'raw_text' (str)
        or None if no metadata found
    """
    try:
        with Image.open(image_path) as img:
            info = img.info

            # For JPEGs, check EXIF data (especially Civitai images)
            exif_comment = None
            try:
                exif = img.getexif()
                if exif:
                    # Try to get EXIF IFD which contains UserComment
                    ifd = exif.get_ifd(0x8769)
                    if ifd and 37510 in ifd:  # 37510 = UserComment
                        raw_bytes = ifd[37510]
                        # Civitai stores metadata as UTF-16-LE with null bytes
                        # Skip UNICODE prefix and strip nulls
                        if raw_bytes.startswith(b'UNICODE\x00'):
                            raw_bytes = raw_bytes[8:]
                        # Strip all null bytes then decode as UTF-8
                        exif_comment = raw_bytes.replace(b'\x00', b'').decode('utf-8', errors='ignore')
            except:
                pass

            # Check for NovelAI format (PNG)
            if "Software" in info and info["Software"] == "NovelAI" and "Comment" in info:
                comment = sanitize_text(info["Comment"], 50000)
                metadata = MetadataNovelAI(comment)
                return {
                    "format": "NovelAI",
                    "params": metadata.params,
                    "raw_text": metadata.raw_text
                }

            # Check for A1111 format in PNG
            if "parameters" in info and isinstance(info["parameters"], str):
                parameters = sanitize_text(info["parameters"], 50000)
                if "Steps:" in parameters:
                    metadata = MetadataA1111(parameters)
                    return {
                        "format": "A1111/Civitai",
                        "params": metadata.params,
                        "raw_text": metadata.raw_text
                    }

            # Check for A1111 format in JPEG EXIF data
            for tag_id, value in exif_data.items():
                if "Steps:" in value:
                    parameters = sanitize_text(value, 50000)
                    metadata = MetadataA1111(parameters)
                    return {
                        "format": "A1111/Civitai (JPEG)",
                        "params": metadata.params,
                        "raw_text": metadata.raw_text
                    }

            # Check for ComfyUI format (PNG)
            if "prompt" in info and isinstance(info["prompt"], str):
                prompt = sanitize_text(info["prompt"], 50000)
                if prompt.lstrip().startswith("{"):
                    workflow = sanitize_text(info.get("workflow", ""), 50000) if "workflow" in info else None
                    metadata = MetadataComfyUI(prompt, workflow)
                    return {
                        "format": "ComfyUI",
                        "params": metadata.params,
                        "raw_text": metadata.raw_text,
                        "workflow_text": metadata.workflow_text
                    }

            # Check for ComfyUI in JPEG EXIF
            for tag_id, value in exif_data.items():
                if value.strip().startswith("{"):
                    prompt = sanitize_text(value, 50000)
                    metadata = MetadataComfyUI(prompt, None)
                    return {
                        "format": "ComfyUI (JPEG)",
                        "params": metadata.params,
                        "raw_text": metadata.raw_text
                    }

            return None

    except Exception as e:
        print(f"Error reading image: {e}", file=sys.stderr)
        return None


def main():
    if len(sys.argv) < 2:
        print("Usage: python parse_image_metadata.py <image_file> [--json]")
        print("       Extracts AI generation metadata from PNG/JPEG images")
        sys.exit(1)

    image_path = sys.argv[1]
    output_json = "--json" in sys.argv

    if not Path(image_path).exists():
        print(f"Error: File not found: {image_path}", file=sys.stderr)
        sys.exit(1)

    metadata = extract_metadata(image_path)

    if not metadata:
        print("No metadata found in image", file=sys.stderr)
        sys.exit(1)

    if output_json:
        # Output as JSON
        output = {
            "format": metadata["format"],
            "params": dict(metadata["params"])
        }
        print(json.dumps(output, indent=2))
    else:
        # Human-readable output
        print(f"Format: {metadata['format']}")
        print("\nParameters:")
        for key, value in metadata["params"].items():
            print(f"\n{key}:")
            print(f"  {value}")


if __name__ == "__main__":
    main()
